discovery_engine_core/web-api/src/bin/ingestion.rs

```diff
@@ -310,78 +335,47 @@ fn serialize_to_ndjson(
     Ok(bytes.freeze())
 }
 
-#[derive(Debug)]
-struct TooManyDocumentsError;
-impl Reject for TooManyDocumentsError {}
+#[derive(Error, Debug, DisplayDoc)]
+enum Error {
+    /// Too many documents send to ingestion system
+    TooManyDocuments,
 
-#[derive(Debug)]
-struct EmbeddingsCalculationError(Vec<String>);
-impl Reject for EmbeddingsCalculationError {}
+    /// Embeddings could not be calculated
+    EmbeddingsCalculation(Vec<DocumentId>),
 
-#[derive(Debug)]
-struct SerializeNdJsonError;
-impl Reject for SerializeNdJsonError {}
+    /// Serialization error
+    SerializeNdJson(GenericError),
 
-#[derive(Debug)]
-struct ElasticOpError;
-impl Reject for ElasticOpError {}
+    /// Upload to elastic did not succeed
+    Elastic(#[source] reqwest::Error),
 
-#[derive(Debug)]
-struct ReceivingOpError;
-impl Reject for ReceivingOpError {}
-
-fn handle_elastic_error(err: reqwest::Error) -> Rejection {
-    error!("ElasticOpError {:#?}", err);
-    warp::reject::custom(ElasticOpError)
+    /// Deserialization of response from elastic instance did not succeed
+    Receiving(#[source] reqwest::Error),
 }
 
-/// An API error serializable to JSON.
-#[derive(Serialize)]
-struct ErrorMessage {
-    code: u16,
-    message: String,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    errored_ids: Option<Vec<String>>,
-}
+impl Reject for Error {}
 
 async fn handle_rejection(err: Rejection) -> Result<impl Reply, Infallible> {
-    let code;
-    let message;
-    let mut errored_ids = None;
-
-    if err.is_not_found() {
-        code = StatusCode::NOT_FOUND;
-        message = "NOT_FOUND";
-    } else if err
-        .find::<warp::filters::body::BodyDeserializeError>()
-        .is_some()
-    {
-        code = StatusCode::BAD_REQUEST;
-        message = "REQUEST_BODY_DESERIALIZATION_ERROR";
-    } else if let Some(ElasticOpError) = err.find() {
-        code = StatusCode::BAD_REQUEST;
-        message = "ELASTIC_ERROR";
-    } else if let Some(ReceivingOpError) = err.find() {
-        code = StatusCode::BAD_REQUEST;
-        message = "RECEIVING_OPERATION_ERROR";
-    } else if let Some(SerializeNdJsonError) = err.find() {
-        code = StatusCode::BAD_REQUEST;
-        message = "NDJSON_SERIALIZATION_ERROR";
-    } else if let Some(EmbeddingsCalculationError(ids)) = err.find() {
-        code = StatusCode::UNPROCESSABLE_ENTITY;
-        message = "UNPROCESSABLE_DOCUMENTS";
-        errored_ids = Some(ids.to_vec());
-    } else {
-        error!("unhandled rejection: {:?}", err);
-        code = StatusCode::INTERNAL_SERVER_ERROR;
-        message = "UNHANDLED_REJECTION";
-    }
-
-    let json = warp::reply::json(&ErrorMessage {
-        code: code.as_u16(),
-        message: message.into(),
-        errored_ids,
-    });
+    let (json, code) = match err.find() {
+        Some(Error::TooManyDocuments) => {
```

Comment:
I think we should return `400` also for `warp::filters::body::BodyDeserializeError`